### PR TITLE
action viewer overlay

### PR DESCRIPTION
**Pull request by sylviameows**
adds a tool tip overlay to code chest menus so you don't have to hover over the reference book to see descriptions. this overlay is toggle able through the config and is disabled by default.

> example2
> example 2: a standard code block with a large description.

> example3
> example 3: example of normal tooltips displaying over the action's description.

> example4
> example 4: example of a call function block displaying it's description. (only works if reference book is in your inventory)

### CODE REVIEW

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+                try {
+                    action = ActionDump.getActionDump().findAction(sign.getFrontText());
+                } catch (Exception ignored) {
+
+                }
+            }
+        }
+    }
+
+    public static ItemStack getReferenceBook() {
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:64]: This should be a utility, not just for this feature. We should also probably construct a hypercube.actiondump.Action from the reference book, which would allow usage of the ref book in the parameters ghosts.
You can just move it outside (to Utility perhaps) but you don't need to construct an action from it

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    public static void onClickChest(BlockHitResult hitResult) {
+        invalidate();
+        var world = CodeClient.MC.world;
+        if (world == null || !Config.getConfig().ActionViewer) return;
+        var position = hitResult.getBlockPos();
+        if (CodeClient.location instanceof Dev dev && dev.isInDev(position) && world.getBlockEntity(position) instanceof ChestBlockEntity) {
+            var signEntity = world.getBlockEntity(position.down().west());
+            if (signEntity instanceof SignBlockEntity sign) {
+                if (sign.getFrontText().getMessage(0, true).contains(Text.literal("CALL FUNCTION"))) {
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:51]: We should use the reference book to cover everything the actiondump doesn't know, so new actions or people who don't have the actiondump at all

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            if (!Config.getConfig().ActionViewer) return null;
+            ItemStack item;
+            if (action == null) {
+                if (!reference) return null;
+                if (book == null) {
+                    item = getReferenceBook();
+                    if (item == null) return null;
+                    var displayBook = item.copy();
+                    var name = displayBook.getName().copy().withColor(0xFFFFFF);
+                    displayBook.setCustomName(name); // continuity with actions
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:98]: Actions have no colour since you don't have an actiondump with colors:
- Running just `/getactiondump` will achieve this.
- Running `/getactiondump section` will get an actiondump with colours, which is intented behaviour.
tl;dr, they are supposed to have color

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        TooltipComponent tooltipComponent;
+        for (Iterator<TooltipComponent> iterator = components.iterator(); iterator.hasNext(); tooltipHeight += tooltipComponent.getHeight()) {
+            tooltipComponent = iterator.next();
+            int width = tooltipComponent.getWidth(textRenderer);
+            if (width > tooltipWidth) {
+                tooltipWidth = width;
+            }
+        }
+        Vector2ic vector = positioner.getPosition(context.getScaledWindowWidth(), context.getScaledWindowHeight(), x, y, tooltipWidth, tooltipHeight);
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:150]: if it's going of the edge, such as a small screen size, we should probably make hovering over the tooltip (or where it normally) is move it into the screen

**Review by GeorgeRNG — CHANGES_REQUESTED** [review]

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java (hunk 1 above)
Comment (sylviameows) [reply src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java]: converting the reference book to an action was going to be my first approach but i hadn't wanted mess with parsing all of that data. however, i have submitted a commit that adds a new class `hypercube.ReferenceBook` that can be used for that purpose in the future with parameters ghosts and also moved `ActionViewer#getReferenceBook()` into `location.Dev` as it felt fitting to go there.

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java (hunk 4 above)
Comment (sylviameows) [reply src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:108]: going to look into this, also do you have any suggestions on how to handle tool tips that are too tall to display on screen? originally i was thinking of doing something with scaling but i couldn't think of a way to do that.

**Comment by sylviameows** [comment]
this feature is much more polished now, and has some extra benefit with viewing code blocks with large action sizes (like noise blocks). I'd like to look into adding some sort of easing on the movements, but that probably wouldn't fit into CodeClient's style.

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        ActionTooltipPositioner.INSTANCE.setMousePosition(mouseX,mouseY); // I would pass these through like normal, but draw tooltip might be moved to a utility class in the future.
+
+        var z = hover ? 500 : 300;
+        drawTooltip(context, handledScreen, transformText(text), 176, scroll, z);
+    }
+
+    public static boolean scroll(Screen screen, double mouseX, double mouseY, double horizontalAmount, double verticalAmount) {
+        if (!(screen instanceof GenericContainerScreen || screen instanceof CustomChestMenu)) return false;
+        if (isValid() && tall && hover) {
+            scroll -= 8 * (int) verticalAmount;
```
Comment (GeorgeRNG) [inline src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:96]: scroll is inverted and can't scroll to the very top or bottom

Comment (GeorgeRNG) [reply src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java:96]: ok scrolling to the top or bottom is a problem with the large custom chest on big enough GUI scales, since that menu goes off the screen. We could ignore this, or account for it.

**Review by GeorgeRNG — CHANGES_REQUESTED** [review]

**Comment by RedVortexDev** [comment]
hey 👋 
this looks really good.. thinking that if the tooltip was pinned to top instead of centered (closer to args) it would look better?

**Comment by sylviameows** [comment]
i've added tooltip alignment and scrolling inversion as preferences in the config, and changed the defaults to have the feature enabled, aligned to the top, and use normal scrolling.

File: src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java (hunk 5 above)
Comment (sylviameows) [reply src/main/java/dev/dfonline/codeclient/dev/overlay/ActionViewer.java]: changed scroll inversion to be a config option (as for me it felt more natural) and fixed the scrolling issue on the large custom chest.
